CDK-988. Add a long range partitioner with fixed size bounds.

tomwhite opened the pull request.

rdblue commented:
@tomwhite, I'd like to do a release soon. Is this something you want to get in?

tomwhite commented:
Thanks for the feedback @laserson and @rdblue! I have updated the PR to use the start value for each bucket in the range.

rdblue commented on kite-data/kite-data-core/src/main/java/org/kitesdk/data/spi/partition/FixedLongRangeFieldPartitioner.java:
```diff
@@ -0,0 +1,120 @@
+      return ((In<Long>) predicate).transform(this);
+    } else if (predicate instanceof Range) {
+      // must use a closed range:
+      //   if this( 5 ) => 10 then this( 6 ) => 10, so 10 must be included
```
This projection method is correct, but there should be a test for it. Also, could you update or remove the comment on why it must be a closed range?

rdblue commented on kite-data/kite-data-core/src/main/java/org/kitesdk/data/spi/partition/FixedLongRangeFieldPartitioner.java:
```diff
@@ -0,0 +1,120 @@
+    if (predicate instanceof Exists) {
+      return Predicates.exists();
+    } else if (predicate instanceof In) {
+      return ((In<Long>) predicate).transform(this);
```
I think this strict projection is incorrect. The only way a value should be matched here is if all of the inputs to this partitioner that produce this value are matched by the predicate. If the partitioner has size 5 and its strict predicate returns 10, then the original predicate must have matched 10, 11, 12, 13, and 14: all of the values that get mapped to 10. Because these are integers, we can actually check them, though returning null is fine with me because it is a pain to check all possible values. Returning null will make Kite apply the original predicate when reading the data.

On the other hand, ranges are easier to reason about and should be included. You can check each endpoint individually: if `(start % size) == 0` then the projected range should start there, otherwise start at the next bucket. If `((end + 1) % size) == 0` then include the bucket for end, otherwise the one before that. Then construct a closed range from the bucket endpoints and be sure to check that start < end. Hopefully that makes sense!

This method will need a test as well.

rdblue replied:
Here's an implementation for ranges:

```java
      Range<Long> closed = Ranges.adjustClosed(
          (Range<Long>) predicate, DiscreteDomains.longs());
      Long start = null;
      if (closed.hasLowerBound()) {
        if ((closed.lowerEndpoint() % size) == 0) {
          // the entire set of values is included
          start = closed.lowerEndpoint();
        } else {
          // start the predicate at the next value
          start = apply(closed.lowerEndpoint() + size);
        }
      }
      Long end = null;
      if (closed.hasUpperBound()) {
        if (((closed.upperEndpoint() + 1) % size) == 0) {
          // all values are included
          end = apply(closed.upperEndpoint());
        } else {
          // end the predicate at the previous value
          end = apply(closed.upperEndpoint() - size);
        }
      }
      if (start != null && end != null && start < end) {
        return null;
      }
      return Ranges.closed(start, end); // null start or end => unbound
```

rdblue commented on kite-data/kite-data-core/src/test/java/org/kitesdk/data/spi/partition/TestFixedLongRangeFieldPartitioner.java:
```diff
@@ -0,0 +1,43 @@
+
+  @Test(expected = IllegalArgumentException.class)
+  public void testRangeNotPositive() {
+    new FixedLongRangeFieldPartitioner("position", 0);
```
Include a negative number as well?

rdblue commented on kite-data/kite-data-core/src/main/java/org/kitesdk/data/PartitionStrategy.java:
```diff
@@ -368,6 +369,50 @@ public Builder range(String sourceName, String... upperBounds) {
+     * @return An instance of the builder for method chaining.
+     * @see IntRangeFieldPartitioner
+     */
+    public Builder fixedRange(String sourceName, long size) {
```
Should this be `fixedSizeRange`?

rdblue commented on kite-data/kite-data-core/src/main/java/org/kitesdk/data/spi/partition/FixedLongRangeFieldPartitioner.java:
```diff
@@ -0,0 +1,120 @@
+    "SE_COMPARATOR_SHOULD_BE_SERIALIZABLE"},
+    justification="False positive due to generics.")
+@Immutable
+public class FixedLongRangeFieldPartitioner extends FieldPartitioner<Long, Long> {
```
I think a better name would be `LongFixedSizedRangePartitioner`. The function is fixed-size-range or fixed-range, so it seems odd to put the type in the middle.

rdblue commented:
One more note: any idea how to extend this for int, float, and double types? For the 4-byte types we could simply promote them and run the long/double partitioner?

I think we would need a separate partitioner for a double fixed range, which I think will be valuable for geo purposes since lat/long values are usually decimals. To do this we could just add a "type" option to the existing "range" JSON. Just want to hear your thoughts here so we have a plan for adding it later that doesn't conflict with this work.

tomwhite commented:
@rdblue Thanks for all the feedback, and for the guidance on implementing the `project` methods. I've added tests for them now, and addressed all your other points.

Regarding double fixed size range - it should be straightforward to add it in the future with a "type" option as you suggest.

rdblue commented on kite-data/kite-data-core/src/main/java/org/kitesdk/data/spi/predicates/In.java:
```diff
@@ -89,7 +89,7 @@ public boolean apply(@Nullable T test) {
   }
 
-  Set<T> getSet() {
+  public Set<T> getSet() {
```
Minor: there is also a method, `Predicates.asSet(In)`, that exposes the inner set. I think the purpose of that was to mimic the use of `Ranges.asSet(Range)`, which has to be a static method because we allow ranges of objects that aren't `Comparable` but back the `asSet` method with a Guava range that requires a `Comparable` type.

I'm fine with changing to use this method everywhere, since this is more straight-forward. But I'd like to either use `Predicates.asSet(In)` or `In#getSet()`, not both.

tomwhite replied:
I reverted my change and used `Predicates.asSet(In)`

rdblue commented on kite-data/kite-data-core/src/test/java/org/kitesdk/data/spi/partition/TestPartitionerProjection.java:
```diff
@@ -270,6 +270,48 @@ public void run() {
+    Assert.assertEquals(Ranges.closed(0L, 15L),
+        fp.project(Ranges.closed(4L, 15L)));
+    Assert.assertEquals(Ranges.closed(5L, 20L),
+        fp.project(Ranges.openClosed(5L, 21L)));
```
It would be good to have some ranges with just one bound tested with project, too.

tomwhite replied:
Done. Thanks for taking a look @rdblue.

rdblue commented:
A added a couple of minor comments, but this looks good overall.

rdblue commented:
+1

Thanks, Tom!
